File: library/src/main/java/com/alamkanak/weekview/MonthLoader.java
package com.alamkanak.weekview;

import com.alamkanak.weekview.interfaces.WeekViewLoader;

import java.util.Calendar;
import java.util.List;

public class MonthLoader implements WeekViewLoader {

    private com.alamkanak.weekview.interfaces.MonthChangeListener mOnMonthChangeListener;

    public MonthLoader(com.alamkanak.weekview.interfaces.MonthChangeListener listener){
        this.mOnMonthChangeListener = listener;
    }

    @Override
    public double toWeekViewPeriodIndex(Calendar instance){
        return instance.get(Calendar.YEAR) * 12 + instance.get(Calendar.MONTH) + (instance.get(Calendar.DAY_OF_MONTH) - 1) / 30.0;
    }

    @Override
    public List<? extends WeekViewEvent> onLoad(int periodIndex){
        return mOnMonthChangeListener.onMonthChange(periodIndex / 12, periodIndex % 12 + 1);
    }

    public com.alamkanak.weekview.interfaces.MonthChangeListener getOnMonthChangeListener() {
        return mOnMonthChangeListener;
    }

    public void setOnMonthChangeListener(com.alamkanak.weekview.interfaces.MonthChangeListener onMonthChangeListener) {
        this.mOnMonthChangeListener = onMonthChangeListener;
    }

    /**
     * @deprecated  code refractoring
     *              {will be removed in next version} </br>
     *              use {@link com.alamkanak.weekview.interfaces.MonthChangeListener}  instead.
     */
    @Deprecated
    public interface MonthChangeListener extends com.alamkanak.weekview.interfaces.MonthChangeListener{

    }
}

File: library/src/main/java/com/alamkanak/weekview/MonthLoader.kt
package com.alamkanak.weekview

import java.util.*

/**
 * This class is responsible for loading [WeekViewEvent]s into [WeekView]. It can handle
 * both concrete [WeekViewEvent] objects and [WeekViewDisplayable] objects. The latter is
 * an interface that can be implemented in one's actual data class and handles the conversion to a
 * [WeekViewEvent].
 */
internal class MonthLoader<T>(
        var onMonthChangeListener: MonthChangeListener<T>?
) : WeekViewLoader<T> {

    override fun toPeriod(instance: Calendar): Period {
        return Period.fromDate(instance.toLocalDate())
    }

    override fun onLoad(period: Period): List<WeekViewEvent<T>> {
        val (month, year) = period

        val startDate = today()
                .withYear(year)
                .withMonth(month)
                .withDayOfMonth(1)
        val maxDays = startDate.lengthOfMonth()
        val endDate = startDate.withDayOfMonth(maxDays)

        val listener = onMonthChangeListener ?: return emptyList()
        return listener
                .onMonthChange(startDate.toCalendar(), endDate.toCalendar())
                .map { it.toWeekViewEvent() }
    }

}
